usersession/userd/launcher.go

```diff
@@ -170,6 +185,216 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
+			return dbus.MakeFailedError(fmt.Errorf("Supplied environment variable %q is not allowed", e))
+		}
+
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	// XXX: this avoids defunct processes but causes userd to persist
+	// until all children are gone (currently, this is not a problem since
+	// userd is long running once started)
+	go cmd.Wait()
+
+	if cmd.Run() != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot run %q", exec_command))
+	}
+
+	return nil
+}
+
+type fileExists func(string) bool
+
+func existsOnFileSystem(desktopFile string) bool {
+	fileStat, err := os.Stat(desktopFile)
+
+	// We only support files in /var/lib/snapd/desktop/applications and they should
+	// all be regular files.
+	return err == nil && fileStat.Mode().IsRegular()
+}
+
+// findDesktopFile recursively tries each subdirectory that can be formed from the (split) desktop file ID.
+// Per https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id,
+// if desktop entries have dashes in the name ('-'), this could be an indication of subdirectories, so search
+// for those too. Eg, given foo-bar_baz_norf.desktop the following are searched for:
+//   o .../foo-bar_baz-norf.desktop
+//   o .../foo/bar_baz-norf.desktop
+//   o .../foo/bar_baz/norf.desktop
+//   o .../foo-bar_baz/norf.desktop
+// We're not required to diagnose multiple files matching the desktop file ID.
+func findDesktopFile(desktopFile_exists fileExists, base_dir string, splitFileId []string) (string, error) {
+	desktopFile := filepath.Join(base_dir, strings.Join(splitFileId, "-"))
+
+	if desktopFile_exists(desktopFile) {
+		return desktopFile, nil
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID.
+	// Maybe this is overkill: At the time of writing, the only use is in desktopFileIDToFilename() and there
+	// we're only checking dirs.SnapDesktopFilesDir (and not all entries in $XDG_DATA_DIRS).
+	// For dirs.SnapDesktopFilesDir snapd will not create any subdirectories.
+	for i := 1; i != len(splitFileId); i++ {
+		desktopFile, err := findDesktopFile(desktopFile_exists, filepath.Join(base_dir, strings.Join(splitFileId[:i], "-")), splitFileId[i:])
+		if err == nil {
+			return desktopFile, err
+		}
+	}
+
+	return "", fmt.Errorf("could not find desktop file")
+}
+
+// desktopFileIDToFilename determines the path associated with a desktop file ID.
+func desktopFileIDToFilename(desktopFile_exists fileExists, desktopFileID string) (string, error) {
+
+	// OpenDesktopEntryEnv() currently only supports launching snap applications from
+	// desktop files in /var/lib/snapd/desktop/applications and these desktop files are
+	// written by snapd and considered safe for userd to process.
+	// Since we only access /var/lib/snapd/desktop/applications, ignore
+	// https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+	base_dir := dirs.SnapDesktopFilesDir
```

Comment:
With the current iteration we don't want to consider other desktop files:
```c++
	// OpenDesktopEntryEnv() currently only supports launching snap applications from
	// desktop files in /var/lib/snapd/desktop/applications and these desktop files are
	// written by snapd and considered safe for userd to process.
```
We'll discuss other locations in the foreseeable future.

Reply:
Isn't the idea to see whether the desktop ID belongs to a snap?  How can you do that if you don't check if something earlier in the search path matches the ID first?

Reply:
Isn't that going to cause weird errors? The app plugging the interface only gets access to `/var/lib/snapd/desktop/applications` and constructs a desktop ID from what it finds there. It **cannot** check whether there is an earlier match in $XDG_DATA_DIRS, so it has no way to anticipate such a failure.

Reply:
This code is not being run by the sandboxed application: it is part of `snap userd`, which is running unconfined.

Reply:
The client code that provides the ID runs with the plugged interface. That defines the locations that can be searched for `.desktop` files, not the environment variables seen by `snap userd`. On the `snap userd` side we should look in the same places the client sees for consistency.

The client cannot know, and should not have to care, that there is another desktop file (with a duplicate ID, and potentially different content) somewhere it cannot and should not access.

Reply:
The whole point of this routine is to determine whether the desktop file ID is owned by a snap.  If the name is shadowed by something earlier in the search path, then the provided ID is not owned by a snap.  To do otherwise would be inconsistent with the host system's view of desktop file IDs.

Reply:
The whole point of this routine is to identify the desktop file referred to by the supplied desktop file ID.

There is nothing absolute about a desktop file ID: It depends on the context. For example, on a traditional desktop two users could refer to different files by the same ID (if their `$XDG_DATA_HOME` and/or `$XDG_DATA_DIRS` is different).

We know the context in which the desktop file ID belongs and should not be applying a different context.

Reply:
That's my entire point: we need to read those environment variables and check all the directories they list in order to resolve the desktop file ID.  By ignoring this configuration, we will produce different results than every other implementation of this standard.

I'll see if I can get the code working the way I believe it should.

Reply:
I think we're in agreement that both sides of this interaction need to be interpreting the desktop file ID in a consistent way, we just differ in how that is achieved.

I'm taking the view that, with the current implementation, the only place it makes sense to look for desktop files is `/var/lib/snapd/desktop/applications` and that requests cannot refer to files stored elsewhere.

You're taking the view that `$XDG_DATA_HOME` and `$XDG_DATA_DIRS` are required to be set in a way in the requesting snap and in `snap userd` so that desktop file IDs match. And that desktop file IDs which _can_ refer to files stored elsewhere need to be rejected.

In practice, I think the most significant usecase affected is when a user customises a file from `/var/lib/snapd/desktop/applications` and places a modified version in `$XDG_DATA_HOME/applications`. My approach executes the unmodified .desktop, yours rejects the launch. Neither is what the user expects, but you can argue that failing is better than doing the wrong thing.

I still think the current approach is "good enough" to get this basic functionality in place. Another possible way forward is for `snap userd` to offer a way to enumerate and access all the .desktop files it can see.

Reply:
My concern is that the current behaviour could make extending the interface in future difficult. It's a lot easier to turn failure cases into successful cases than to change the behaviour or existing successful cases.

If a snap's desktop file is shadowed by something earlier in the search path, then it would effectively be invisible to the desktop shell on the host system.  I don't think it is it's a bad thing that attempts to launch such a desktop file would fail here.